docs: Migration Guide v3 (#2140)
fastify/fastify: docs/Migration-Guide-V3.md

# V3 Migration Guide

This guide is aimed to help migrating from Fastify v2 to v3.

> Before beginning please ensure that any deprecation warnings from v2 are fixed as we have removed all deprecated features and they will no longer work after upgrading. (#1750)

Reviewer: Please use full URLs for the PRs.

## Breaking changes

### Changed middleware support (#2014)

From Fastify v3, middleware support does not come out of the box with the framework itself.

If you use Express middleware in your application, please install and register the [`fastify-express`](https://github.com/fastify/fastify-express) or [`middie`](https://github.com/fastify/middie) plugin before doing so.

**v2:**

```js
fastify.use(require("cors")());
```

**v3:**

```js
await fastify.register(require("fastify-express"));
fastify.use(require("cors")());
```

### Changed schema substitution (#2023)

> We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant `$ref-way` references.

Reviewer: Please add a link to the documentation of the "replace-way" behavior. Also, `$ref-way` -> `$ref`. The standard is JSON Schema.

**v2:**

```js
const schema = {
  body: "schemaId#"
};
fastify.route({ method, url, schema, handler });
```

**v3:**

```js
const schema = {
  body: {
    $ref: "schemaId#"
  }
};
fastify.route({ method, url, schema, handler });
```

### Changed schema validation options (#2023)

We have replaced `setSchemaCompiler` and `setSchemaResolver` options with `setValidatorCompiler` to enable future tooling improvements.

**v2:**

```js
const fastify = Fastify();
const ajv = new AJV();
ajv.addSchema(fastClone(schemaA));

> ajv.addSchema(fastClone(schemaB));

Reviewer: What is `fastClone`? Is it needed?

Author: Was wondering the same, I took it from the example here: https://github.com/fastify/fastify/pull/2023
@Eomm Can I remove the fastClone or is that something that was required in v2?

fastify.setSchemaCompiler(schema => ajv.compile(schema));
fastify.setSchemaResolver(ref => ajv.getSchema(ref).schema);
```

**v3:**

```js
const fastify = Fastify();
const ajv = new AJV();
ajv.addSchema(schemaA);
ajv.addSchema(schemaB);

fastify.setValidatorCompiler((method, url, httpPart, schema) =>
  ajv.compile(schema)
);

> ```

Reviewer: @Eomm have you got anything to add here?

Reviewer 2: Yes, I would mention the new feature `setSerializerCompiler`

Moreover, I'm writing a blog post about these features so I will add a link to study in deep

### Changed hooks behaviour (#2004)

From Fastify v3, the behavior of `onRoute` and `onRegister` hooks will change slightly in order to support hook encapsulation.

> - `onRoute` - The hook will be called asynchronously, in v1/v2 it's called as soon as a route is registered. This means that if you want to use it, you should register this hook as soon as possible in your code.

Reviewer: I don't think there is `onRoute` in v1.

Author: I've just realised that the documentation for this needs to be updated in `Hooks.md` - It's still talking about onRoute being trigger synchronously:
```
### onRoute
Triggered when a new route is registered. Listeners are passed a `routeOptions` object as the sole parameter. The interface is synchronous, and, as such, the listeners do not get passed a callback. This hook is encapsulated.
```

Reviewer: Would you mind to send a PR to update the docs?

Author: I think I misunderstood the sentence. What the author meant is probably that the hook runs synchronously (fire and forget) which isn't a contradiction to it being triggered asynchronously during route registration. 

I'm not clear exactly what changed though and what users need to be aware of here. 
@Eomm do you have more context of this change? Are the docs correct or can you clarify the change so I can update it?

Reviewer: > I think I misunderstood the sentence. What the author meant is probably that the hook runs synchronously (fire and forget) which isn't a contradiction to it being triggered asynchronously during route registration.

Exactly.

---

The key difference between v2 and v3 in this case is that onRoute hooks will be called only for routes nested in that plugin and descendants. In case of hooks defined in something wrapped in `fastify-plugin`, the hook will be registered in its current context. 

---

The difference is

- `onRegister` - Same as the onRoute hook, the only difference is that now the very first call will no longer be the framework itself, but the first registered plugin

### Changed TypeScript support

TODO: Our types have been completely rewritten

## Further additions and improvements

- Hooks now have consistent context irregardless of how they are registered (#2005)
- Deprecated `request.req` and `reply.res` for [`request.raw`](Request.md) and [`reply.raw`](Reply.md) (#2008)
- Removed `modifyCoreObjects` option (#2015)

> - [Serializers](Logging.md) now receive Fastify request and response objects instead of native ones (#2017)

Reviewer: This is important as well as it's something that can break several apps (most customize the logger).

- Added [`connectionTimeout`](Server.md#factory-connection-timeout) option (#2086)
- Added [`keepAliveTimeout`](Server.md#factory-keep-alive-timeout) option (#2086)

> - Added async-await support for [plugins](Plugins.md#async-await) (#2093)

Reviewer: This is a big deal and it deserver its own section.

Author: Is it a breaking change though? The migration guide should be focussed on helping users upgrade to latest version.

# V3 Migration Guide

This guide is aimed to help migration from Fastify v2 to v3.

Before beginning please ensure that any deprecation warnings from v2 are fixed as we have removed all deprecated features and they will no longer work after upgrading. ([#1750](https://github.com/fastify/fastify/pull/1750))

## Breaking changes

### Changed middleware support ([#2014](https://github.com/fastify/fastify/pull/2014))

From Fastify v3, middleware support does not come out of the box with the framework itself.

If you use Express middleware in your application, please install and register the [`fastify-express`](https://github.com/fastify/fastify-express) or [`middie`](https://github.com/fastify/middie) plugin before doing so.

**v2:**

```js
fastify.use(require("cors")());
```

**v3:**

```js
await fastify.register(require("fastify-express"));
fastify.use(require("cors")());
```

> ### Changed logging serialisation ([#2017](https://github.com/fastify/fastify/pull/2017))

Reviewer 2: ```suggestion
### Changed logging serialization ([#2017](https://github.com/fastify/fastify/pull/2017))
```

```

**v3:**

```js
await fastify.register(require('fastify-express'));
fastify.use(require('cors')());
```

### Changed logging serialization ([#2017](https://github.com/fastify/fastify/pull/2017))

We have updated our logging [Serializers](https://github.com/fastify/fastify/blob/master/docs/Logging.md) to now receive Fastify [`Request`](https://github.com/fastify/fastify/blob/master/docs/Request.md) and [`Reply`](https://github.com/fastify/fastify/blob/master/docs/Reply.md) objects instead of native ones.

If you have created custom serializers they will need updating if they expect properties that aren't exposed by the Fastify objects themselves.

**v2:**

```js
const fastify = require('fastify')({
  logger: {
    serializers: {
      res(res) {
        return {
          statusCode: res.statusCode,
          customProp: res.customProp
        };
      }
    }
  }
});
```

**v3:**

```js
const fastify = require('fastify')({
  logger: {
    serializers: {
      res(reply) {
        return {
          statusCode: reply.statusCode, // No change required
          customProp: reply.raw.customProp // Log custom property from res object
        };
      }
    }
  }
});
```

> ### Changed schema substitution ([#2023](https://github.com/fastify/fastify/pull/2023))

Reviewer: We also used to support `@hapi/joi`. I think the schema-replacement functionality was helping Joi to reuse schemas. As a result, we do not support this anymore. Can you investigate and add a paragraph about it?

Author: This is still working but I've fixed the code snippet in the docs

> We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant JSON Schema `$ref` based substitution.

Reviewer 2: ```suggestion
We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant JSON Schema `$ref` based substitution. To deepen this change [read the article](https://dev.to/eomm/validation-and-serialization-in-fastify-v3-2e8l).
```

**v2:**

```js
const schema = {
  body: 'schemaId#'
};
fastify.route({ method, url, schema, handler });
```

**v3:**

```js
const schema = {
  body: {
    $ref: 'schemaId#'
  }
};
fastify.route({ method, url, schema, handler });
```

### Changed schema validation options ([#2023](https://github.com/fastify/fastify/pull/2023))

> We have replaced `setSchemaCompiler` and `setSchemaResolver` options with `setValidatorCompiler` to enable future tooling improvements.

Reviewer 2: ```suggestion
We have replaced `setSchemaCompiler` and `setSchemaResolver` options with `setValidatorCompiler` to enable future tooling improvements. To deepen this change [read the article](https://dev.to/eomm/validation-and-serialization-in-fastify-v3-2e8l).
```

**v2:**

```js
const fastify = Fastify();
const ajv = new AJV();
ajv.addSchema(schemaA);
ajv.addSchema(schemaB);

fastify.setSchemaCompiler(schema => ajv.compile(schema));
fastify.setSchemaResolver(ref => ajv.getSchema(ref).schema);
```

**v3:**

```js
const fastify = Fastify();
const ajv = new AJV();
ajv.addSchema(schemaA);
ajv.addSchema(schemaB);

> fastify.setValidatorCompiler((method, url, httpPart, schema) =>

Reviewer 3: This is no longer correct after #2179. There should also be a significant call out to all of the signature changes in that PR.

Reviewer 3: Not everything about this is resolved. Please refrain from marking things resolved before the reviewer has a chance to say they are satisfied.

What is still missing is a sentence or two that specifically calls attention to the fact that the function signatures have changed.

### Changed schema substitution ([#2023](https://github.com/fastify/fastify/pull/2023))

> We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant JSON Schema `$ref` based substitution. To deepen this change [read the article](https://dev.to/eomm/validation-and-serialization-in-fastify-v3-2e8l).

Reviewer 3: ```suggestion
We have dropped support for non-standard `replace-way` shared schema substitution and replaced it with standard compliant JSON Schema `$ref` based substitution. To better understand this change [read Validation and Serialization in Fastify v3](https://dev.to/eomm/validation-and-serialization-in-fastify-v3-2e8l).
```
